File: typescript/src/day1.ts
import { readInput, day, mockLog } from './helpers';

function part1(input) {
  console.log(reduce(input));
}

function part2(input) {
  const threeSums = input
    .map((depth, index, array) => {
      if (index + 2 < array.length - 1) return depth + array[index + 1] + array[index + 2];
    })
    .filter(x => x);

  console.log(reduce(threeSums));
}

const reduce = depths =>
  depths.reduce((acc, depth, index, array) => {
    acc += Number(depth > array[index - 1]);
    return acc;
  }, 0);

if (require.main === module) {
  const input = readInput(day(__filename)).map(x => parseInt(x));

  if (process.argv[2] == '1') part1(input);
  else if (process.argv[2] == '2') part2(input);
  else {
    console.log(`day ${day(__filename)}`);
    console.time('part1');
    part1(input);
    console.timeEnd('part1');
    console.time('part2');
    part2(input);
    console.timeEnd('part2');
  }
} else {
  const sample = `199
200
208
210
200
207
240
269
260
263`.split('\n');

  test('part1', () => {
    const logSpy = mockLog();
    part1(sample);
    expect(logSpy).toBeCalledWith(7);
  });

  test('part2', () => {
    const logSpy = mockLog();
    part2(sample);
    expect(logSpy).toBeCalledWith(5);
  });
}

File: rust/src/day1.rs
use std::num::ParseIntError;

use aoc_runner_derive::{aoc, aoc_generator};

#[aoc_generator(day1)]
pub fn generator(input: &str) -> Result<Vec<u32>, ParseIntError> {
    input.lines().map(str::parse).collect()
}

pub fn reduce(input: &Vec<u32>) -> u32 {
    return input.iter().enumerate().fold(0, |acc, (index, depth)| {
        acc + (depth > &input[if index == 0 { 0 } else { index - 1 }]) as u32
    });
}

#[aoc(day1, part1)]
pub fn part1(input: &Vec<u32>) -> u32 {
    reduce(input)
}

#[aoc(day1, part2)]
pub fn part2(input: &Vec<u32>) -> u32 {
    let three_sums = input
        .iter()
        .enumerate()
        .map(|(index, depth)| {
            if index + 2 < input.len() - 1 {
                depth + input[index + 1] + input[index + 2]
            } else {
                0
            }
        })
        .filter(|x| x > &0)
        .collect();

    reduce(&three_sums)
}

#[cfg(test)]
mod tests {
    use super::*;

    const SAMPLE: &str = "199
200
208
210
200
207
240
269
260
263";

    #[test]
    fn sample1() {
        assert_eq!(part1(&generator(SAMPLE).unwrap()), 7);
    }

    #[test]
    fn sample2() {
        assert_eq!(part2(&generator(SAMPLE).unwrap()), 5);
    }
}
